### PR TITLE
Add visualization server.

### LINKED ISSUE #227: Add filler callback with future support
It might be nice to have a version of this https://github.com/NSLS-II/bluesky/pull/1099/files which a) returns documents b) takes in an optional executor which reports futures rather than data enabling parallel data loading (looking at you dask).

### LINKED ISSUE #229: Copy `RunRouter`
It would be good to have a tested copy of `bluesky` `RunRouter`. This way we can make certain that the new pipeline generation code will work for next cycle. (This also means we can start building things on it before it is `bluesky` released)

### PR DESCRIPTION
**Pull request by CJ-Wright**
This server will handle any and all data visualization needs and can scale
to as many or as few figures as needed. This will remove pressure on the
analysis server as it won't need to support both the data processing
and data visualization.

Closes https://github.com/xpdAcq/mission-control/issues/158
Closes #229 
Closes #227

### CODE REVIEW

**Comment by CJ-Wright** [comment]
This needs support for waterfall plots.

**Comment by CJ-Wright** [comment]
I might merge this so we have some of the callbacks into the codebase. This way we can work on a bunch of the servers in parallel (tomo, saving, etc.)

**Comment by sbillinge** [comment]
I can give a quick review right now if you are ready to merge....

**Comment by CJ-Wright** [comment]
That would be great! Sorry that the PR is so huge!

File: xpdan/vend/callbacks/broker.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
+import os
+import time as ttime
+from .core import CallbackBase, Retrieve
+from bluesky.utils import ensure_uid
+import numpy as np
+
+
+class BrokerCallbackBase(CallbackBase):
+    """
+    Base class for callbacks which need filled documents
```
Comment (sbillinge) [inline xpdan/vend/callbacks/broker.py]: I think this needs a tad more explanation in a longer para below....for example, what is a "filled document"?  Why does the base class need this?  Is this broker base-class filling that need?  If so, it needs to say something like "extension to Callbackbase to solve such and such problem"

File: xpdan/vend/callbacks/broker.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+            res, = self.db.fill_events(
+                events=[doc],
+                descriptors=[self.descriptor_dict[doc['descriptor']]],
+                fields=fields)
+            doc['data'].update(**res['data'])  # modify in place
+
+
+class LiveImage(CallbackBase):
+    """
+    Stream 2D images in a cross-section viewer.
```
Comment (sbillinge) [inline xpdan/vend/callbacks/broker.py]: likewise, please take a moment to try and see this comment from the point of view of someone who is not you.  What does this class really do?  Who would use it and why would they use it?  This is then a much more usef docstring moving forward

File: xpdan/vend/callbacks/broker.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+        # use our copy instead of expecting the header to include one.
+        if name != 'stop':
+            callback('stop', doc)
+
+    return f
+
+
+def make_restreamer(callback, db):
+    """
+    Run a callback whenever a uid is updated.
```
Comment (sbillinge) [inline xpdan/vend/callbacks/broker.py]: likewise, see above

File: xpdan/startup/viz_server.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+plt.ion()
+
+d = RemoteDispatcher(glbl_dict["outbound_proxy_address"])
+install_qt_kicker(loop=d.loop)  # This may need to be d._loop depending on tag
+
+figure_pool = {}
+
+
+def fig_factory(x):
```
[the remark this answers is not in the document]
Comment (CJ-Wright) [reply xpdan/startup/viz_server.py]: It shouldn't be user facing, but might as well

**Comment by CJ-Wright** [comment]
Although I am very sympathetic to the docs concerns I think that we should push on with this and perform a large doc-a-thon before release. This will give us the opportunity to move forward (which may actually get rid of some code (lookin at you `BrokerCallbackBase`)) and then summarize our work once we feel ready to release it.

**Pull request merged**